chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,197 @@
+    package "ceph-common" do
+      action :nothing
+    end.run_action(:install)
+  end
+end
+
+if node[:platform] == "suse"
+  package "qemu-block-rbd" do
+    action :install
+    only_if { node[:platform_version].to_f >= 12.0 }
+  end
+end
+
+# Second loop to do our setup
+cinder_controller[:cinder][:volumes].each_with_index do |volume, volid|
+  next unless volume[:backend_driver] == "rbd"
+
+  rbd_user = volume[:rbd][:user]
+  rbd_uuid = volume[:rbd][:secret_uuid]
+
+  if volume[:rbd][:use_crowbar]
+    ceph_conf = "/etc/ceph/ceph.conf"
+    admin_keyring = "/etc/ceph/ceph.client.admin.keyring"
+  else
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
+
+    if ceph_conf.empty? || !File.exists?(ceph_conf)
+      Chef::Log.info("Ceph configuration file is missing; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+
+    if !admin_keyring.empty? && File.exists?(admin_keyring)
+      cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s"]
+      check_ceph = Mixlib::ShellOut.new(cmd)
+
+      unless check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
+        Chef::Log.info("Ceph cluster is not healthy; skipping the ceph setup for backend #{volume[:backend_name]}")
+        next
+      end
+    else
+      # Check if rbd keyring was uploaded manually by user
+      client_keyring = "/etc/ceph/ceph.client.#{rbd_user}.keyring"
+      unless File.exists?(client_keyring)
+        Chef::Log.info("Ceph user keyring wasn't provided for backend #{volume[:backend_name]}")
+        next
+      end
+    end
+
+  end
+
+  secret_file_path = "/etc/ceph/ceph-secret-#{rbd_uuid}.xml"
+  
+  file secret_file_path do
+    owner "root"
+    group "root"
+    mode "0640"
+    content "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
+  end #file secret_file_path
+
+  ruby_block "save nova key as libvirt secret" do
+    block do
+      # Check if libvirt is installed and started
+      if system("virsh hostname &> /dev/null")
+
+        # First remove conflicting secrets due to same usage name
+        virsh_secret = Mixlib::ShellOut.new("virsh secret-list")
+        secret_list = virsh_secret.run_command.stdout
+        virsh_secret.error!
+
+        secret_lines = secret_list.strip.split("\n")
+        if secret_lines.length < 2 || !secret_lines[0].start_with?("UUID") || !secret_lines[1].start_with?("----")
+          raise "cannot fetch list of libvirt secret"
+        end
+        secret_lines.shift(2)
+
+        secret_lines.each do |secret_line|
+          secret_uuid = secret_line.split(" ")[0]
+          cmd = ["virsh", "secret-dumpxml", secret_uuid]
+          virsh_secret_dumpxml = Mixlib::ShellOut.new(cmd)
+          secret_xml = virsh_secret_dumpxml.run_command.stdout
+          virsh_secret_dumpxml.error!
```

Comment:
Technically, I wouldn't fail here in case of an error since I could imagine a race where the secret got removed between the secret-list and this. It doesn't matter too much, I guess, but I would just log some message and do a `next`